Add plain='full' option to TextInput

jcfilben opened the pull request:
#### What does this PR do?
Previously on TextInput when the plain prop was set there was still 11px of padding. Demonstrated in this codesandbox: https://codesandbox.io/s/textinput-plain-issue-forked-w4t6q?file=/index.js:285-333

Changing plain to have `padding: 0` would cause backwards compatibility issues. The most notable ones being on Form when there is a textInput. Notice in the screenshot below how the padding was affected around some of the fields ('initial', 'initial@my.com', and 'large') and the Form doesn't look right with these changes. When TextInput is contained in a form the plain prop is set to true, which is why there are so many snapshot changes for Form with this change.
<img width="439" alt="Screen Shot 2020-09-09 at 11 07 09 AM" src="https://user-images.githubusercontent.com/54560994/92630155-a1ac7a80-f28c-11ea-8f81-c3bcbe5856b8.png">

**previously the form looked like this:**

<img width="429" alt="Screen Shot 2020-09-09 at 11 18 44 AM" src="https://user-images.githubusercontent.com/54560994/92631176-3b285c00-f28e-11ea-98e1-31a8c34b8845.png">

To prevent backwards compatibility issues, plain='full' was added as an option to remove the padding. Plain={true} maintains its original behavior.

#### What testing has been done on this PR?
storybook testing with the following code:
```javascript
import React from 'react';
import { storiesOf } from '@storybook/react';

import { Grommet, TextInput } from 'grommet';
import { grommet } from 'grommet/themes';

const SimpleTextInput = () => {
  return (
    <Grommet full theme={grommet}>
          <TextInput
            plain='full'
            placeholder="Its plain, should not have padding"
            name="name"
            type="text"
          />
    </Grommet>
  );
};

storiesOf('TextInput', module).add('TEST plain', () => <SimpleTextInput />, {
  chromatic: { disable: true },
});

```

#### What are the relevant issues?
Closes #4078 

#### Do the grommet docs need to be updated?
Yes, updated in this PR

#### Should this PR be mentioned in the release notes?
Yes

#### Is this change backwards compatible or is it a breaking change?
Backwards compatible

ShimiSun commented on src/js/utils/styles.js:
```diff
@@ -458,4 +458,5 @@ export const sizeStyle = (name, value, theme) => css`
 export const plainInputStyle = css`
   outline: none;
   border: none;
+  padding: 0px;
```
```suggestion
  padding: 0;
```

ShimiSun reviewed (commented):
I understand your concerns with this PR, and I need to think about whether we should apply it and how to communicate it on the release notes.

taysea commented:
To avoid these backwards compatibility issues, would it make sense in FormField to retroactively add back the padding to TextInput that is being taken away by `plain`?

jcfilben commented:
@taysea I think that is definitely a possible solution. The person who originally filed an issue for this #4078 was using textInput within a FormField and they wanted to be able to have the plain option for textInput. So maybe when you do
```javascript
<FormField>
  <TextInput>
</FormField>
```
It would add the padding like you mentioned but if you did
```javascript
<FormField>
  <TextInput plain>
</FormField>
```
It would not have the padding

taysea commented:
@jcfilben Gotcha. Yeah this is tricky because it seems like Grommet automatically applies `plain` to input components when they're in a form (https://github.com/grommet/grommet/blob/master/src/js/components/FormField/FormField.js#L98). Now that we're stripping out the padding with your changes (which I think makes sense conceptually to align with the docs definition), we're seeing a misalignment with the rest of the form text / inputs. The issue of aligning with the docs here is that it feels like we are going to be introducing lots of visual regressions for users.

I'm trying to think about all the changes that might result for people who are currently using `<TextInput plain />` based on this, and it feels difficult to be able to anticipate all the ways grommet users may be using the component.

Another thought around these changes:
- What if I just want to remove the border for a single TextInput (not style my entire theme to have no border) -- how would I do this via grommet or would I need to rely on `style={{ ... }}` which we typically try to avoid?

I know, for example, we are using a `plain` TextInput for the search input on https://design-system.hpe.design/ and we wouldn't want to lost the padding here. Just something to consider around not taking away functionality from users.

ericsoderberghp commented:
I agree with the comments that we can't make this change due to the compatibility issues. I can see a couple of ways to address the issue:
- connect it to something in the theme such that the `plain` prop removes padding, such as `theme.textInput.plain = 'full';`
- add another value, such as `plain='full'`

ShimiSun commented:
I'm OK with proceeding with the suggestion above of
> add another value, such as plain='full'

jcfilben commented:
**Update:** I went ahead with Eric's suggestion and added plain='full'. I added snapshot tests to TextInput to test that plain='full' removes padding and plain={true} still has padding. I updated the docs to included explanation for plain='full'.

This change fixes the compatibility issue, the snapshots are no longer being affected and I think it is now safe to merge

ShimiSun commented on src/js/components/TextInput/README.md:
```diff
@@ -183,10 +183,12 @@ node
 
 Whether this is a plain input with no border or padding.
-      Only use this when the containing context provides sufficient affordance
+      Only use this when the containing context provides sufficient affordance.
```
I think the first line isn't accurate if it's referring to the boolean value
> Whether this is a plain input with no border or padding.
Should it be something like:
```
Whether this is a plain input with no border or outline.
Only use this when the containing context provides sufficient affordance.
Use "full" to remove the padding in addition to removing the border and outline.
```
Thoughts?

ShimiSun replied:
@jcfilben where are we with this comment?

jcfilben replied:
I updated the readme to say 'no border or outline.' instead of 'no border or padding'

ShimiSun commented on src/js/components/TextInput/doc.js:
```diff
@@ -1,5 +1,6 @@
 import { describe, PropTypes } from 'react-desc';
 
+import { plainPropType } from '../../utils/prop-types';
```
The `plain="full"` is very specific for TextInput and not something we would like to encourage in other components (with the thought of keeping the API simple), hence there is no need to extract the logic to the utils file.

jcfilben replied:
Moved plainPropType to just be in TextInput

ShimiSun commented on src/js/utils/styles.js:
```diff
@@ -459,3 +459,14 @@ export const plainInputStyle = css`
 `;
+
+export const getPlainStyle = plain => {
+  if (plain === 'full') {
```
Along the same lines, this behavior is very specific to TextInput hence the logic should be placed TextInput.

jcfilben replied:
I moved getPlainStyle to just be in TextInput

ShimiSun commented on src/js/components/TextInput/StyledTextInput.js:
```diff
@@ -9,9 +10,25 @@ import {
 
+const getPlainStyle = plain => {
+  if (plain === 'full') {
+    return css`
+      outline: none;
+      border: none;
+      padding: 0;
+    `;
```
we should leverage the existing `plainInputStyle` instead of specifying a separate css.

jcfilben replied:
Updated to use `plainInputStyle`
```javascript
return css`
      ${plainInputStyle} padding: 0;
    `;
```

ShimiSun reviewed (commented):
We should update the *.d.ts (typescript file of each component) to include the `full` value of the plain prop.

jcfilben commented:
Update: added plain: full option to index.d.ts. Github is showing lots of changes for this file, this is because of prettier. Only real change was line 32 `plain?: boolean | 'full';`

ShimiSun commented on src/js/components/TextInput/StyledTextInput.js:
```diff
@@ -9,9 +10,23 @@ import {
+  return plain && plainInputStyle;
+};
+
+const plainPropType = PropTypes.oneOfType([
+  PropTypes.bool,
+  PropTypes.oneOf(['full']),
+]);
+
```
Why are propTypes defined on the Styled file? this is not consistent with other components, it is usually kept on the doc.js file.

jcfilben replied:
Moved propTypes to doc.js in latest push

ShimiSun commented on src/js/components/TextInput/doc.js:
```diff
@@ -93,9 +94,10 @@ export const doc = TextInput => {
-      `Whether this is a plain input with no border or padding.
-      Only use this when the containing context provides sufficient affordance`,
+    plain: plainPropType.description(
+      `Whether this is a plain input with no border or outline.
+      Only use this when the containing context provides sufficient affordance.
+      Use "full" to remove padding in addition to removing border and outline.`,
```
I think the context will make more sense if lines 2 & 3 could be flipped.

jcfilben replied:
Switched these lines in latest push

ShimiSun commented on src/js/components/TextInput/__tests__/__snapshots__/TextInput-test.js.snap:
```diff
@@ -2640,6 +2717,84 @@ exports[`TextInput should not have accessibility violations 1`] = `
+  background: transparent;
+  color: inherit;
+  width: 100%;
+  padding: 11px;
```
How come the snapshot shows the TextInput still have padding when plain is full.

jcfilben replied:
The padding is originally set to 11px but then modified to be 0. Look at line 2745 in the snapshot: `padding: 0;`
